README
REQUIREMENTS

This package only needs a C compiler. If you succeded to install
Apertium before, you have it.


DESCRIPTION

This package countains à series of C programs used to process different
format of document, before translating them with Apertium.

Presently, this package include tools to translate :

- man pages
- mnemonic interface files


FILES

The main files of this package are :

desman.c       : source program of the man pages deformatter
desmnemo.c     : source program of the mnemonic interface file deformatter
reman.c        : source program of the man pages reformatter
remnemo.c      : source program of the mnemonic interface file reformatter
apertium-man   : a shell to make man page translations easier
apertium-mnemo : a shell to make mnemonic interface file translations easier
makefile       : makefile to compile and install tools


DIRECTORIES

man            : man pages of the package (presently in English and French)
samples-mnemo  : examples of mnemonic interface files UTF-8 encoded.


NEWS
13/06/18 : Now, desmnemo ans resmnemo commands allow correct translations
(no interaction between two following lines) with mnemonic files in which
the right part string is not included into " "


reman.c
/*
 * Copyleft Bernard Chardonneau (2012-+)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *
 * Generic reformater for mnemonic interface files and man pages
 * after an Apertium translation.
 *
 * This source file allows to generate apertium-reman executable.
 *
 */

#include <stdio.h>
#include <string.h>


/*
 *  main and only function
 */
int main (int argc, char **argv)
{
    int lastchar;   // last character read on input. Must be an int to test EOF


    // if command-line arguments used
    if (--argc >= 1)
    {
        // display help if asked or if too many arguments
        if (strcmp (argv [1], "-h") == 0 || strcmp (argv [1], "--help") == 0 || argc > 2)
        {
            printf ("USAGE: %s [input_file [output_file]\n", argv [0]);
            printf ("  or : %s - [output_file]\n", argv [0]);
            printf ("generic format processor\n");
            return (0);
        }

        // input from a file if needed
        if (strcmp (argv [1], "-") != 0)
        {
            if (! freopen (argv [1], "r", stdin))
            {
                fprintf (stderr, "Cannot access to file %s\n", argv [1]);
                return (1);
            }
        }

        // output into a file if needed
        if (argc >= 2)
        {
            if (! freopen (argv [2], "w", stdout))
            {
                fprintf (stderr, "Cannot write result to file %s\n", argv [2]);
                return (2);
            }
        }
    }

    // beginning
    lastchar = getchar ();

    // while input was not read completly
    while (lastchar != EOF)
    {
        // if '.' found
        // skip completly ".[]" strings but output the '.' in the other cases
        if (lastchar == '.')
        {
            lastchar = getchar ();

            if (lastchar == '[')
            {
                lastchar = getchar ();

                if (lastchar != ']')
                    putchar ('.');
            }
            else
                putchar ('.');
        }
        // other characters
        else
        {
            // there may be plenty of more complicated solutions
            // but this simple one has my preference
            switch (lastchar)
            {
                // skip [ and ]
                case '[' :
                case ']' : break;

                // skip \ but always output the next character
                case '\\': lastchar = getchar ();
                           putchar (lastchar);
                           break;

                // output other characters
                default  : putchar (lastchar);
                           break;
            }

            // read the next character
            lastchar = getchar ();
        }
    }

    // end of file
    return (0);
}


remnemo.c
/*
 * Copyleft Bernard Chardonneau (2012-+)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *
 * Generic reformater for mnemonic interface files and man pages
 * after an Apertium translation.
 *
 * This source file allows to generate apertium-remnemo executable.
 *
 */

#include <stdio.h>
#include <string.h>


/*
 *  main and only function
 */
int main (int argc, char **argv)
{
    int lastchar;   // last character read on input. Must be an int to test EOF


    // if command-line arguments used
    if (--argc >= 1)
    {
        // display help if asked or if too many arguments
        if (strcmp (argv [1], "-h") == 0 || strcmp (argv [1], "--help") == 0 || argc > 2)
        {
            printf ("USAGE: %s [input_file [output_file]\n", argv [0]);
            printf ("  or : %s - [output_file]\n", argv [0]);
            printf ("generic format processor\n");
            return (0);
        }

        // input from a file if needed
        if (strcmp (argv [1], "-") != 0)
        {
            if (! freopen (argv [1], "r", stdin))
            {
                fprintf (stderr, "Cannot access to file %s\n", argv [1]);
                return (1);
            }
        }

        // output into a file if needed
        if (argc >= 2)
        {
            if (! freopen (argv [2], "w", stdout))
            {
                fprintf (stderr, "Cannot write result to file %s\n", argv [2]);
                return (2);
            }
        }
    }

    // beginning
    lastchar = getchar ();

    // while input was not read completly
    while (lastchar != EOF)
    {
        // if ',' found
        // skip completly ",[]" strings but output the ',' in the other cases
        if (lastchar == ',')
        {
            lastchar = getchar ();

            if (lastchar == '[')
            {
                lastchar = getchar ();

                if (lastchar != ']')
                    putchar (',');
            }
            else
                putchar (',');
        }
        // other characters
        else
        {
            // there may be plenty of more complicated solutions
            // but this simple one has my preference
            switch (lastchar)
            {
                // skip [ and ]
                case '[' :
                case ']' : break;

                // skip \ but always output the next character
                case '\\': lastchar = getchar ();
                           putchar (lastchar);
                           break;

                // output other characters
                default  : putchar (lastchar);
                           break;
            }

            // read the next character
            lastchar = getchar ();
        }
    }

    // end of file
    return (0);
}


desmnemo.c
/*
 * Copyleft Bernard Chardonneau (2012-+)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *
 * Deformater for mnemonic interface files before Apertium translation.
 *
 * Files to be used are made of lines with the following formats :
 *
 * Mnemonic             "printf_formated_string"
 *            or now
 * Mnemonic             formated_string
 *
 * The mnemonic starts at colomn 1 and only the printf_formated_string
 * or the formatted_string must be translated by Apertium.
 *
 * Generally, the string on the right is just ordinary text, but %s
 * %d or a similar indication can be included inside if the value of
 * a variable must be included in the generated message.
 *
 * Comments starting by a # at column 1 are also allowed.
 *
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>


/*
 *  returns 1 if a \ needs to be inserted before the character
 *  0 otherwise
 */
int char_to_escape (int caract)
{
    switch (caract)
    {
        case '[' :
        case ']' :
        case '<' :
        case '>' :
        case '{' :
        case '}' :
        case '\\':
        case '/' :
        case '^' :
        case '$' :
        case '@' : return 1;
                   // no break necessary here

        default  : return 0;
    }
}


/*
 *  main function
 */
int main (int argc, char **argv)
{
    int lastchar;  // last character read on input. Must be an int to test EOF


    // if command-line arguments used
    if (--argc >= 1)
    {
        // display help if asked or if too many arguments
        if (strcmp (argv [1], "-h") == 0 || strcmp (argv [1], "--help") == 0 || argc > 2)
        {
            printf ("USAGE: %s [input_file [output_file]\n", argv [0]);
            printf ("  or : %s - [output_file]\n", argv [0]);
            printf ("mnemonic interface file format processor\n");
            return (0);
        }

        // input from a file if needed
        if (strcmp (argv [1], "-") != 0)
        {
            if (! freopen (argv [1], "r", stdin))
            {
                fprintf (stderr, "Cannot access to file %s\n", argv [1]);
                return (1);
            }
        }

        // output into a file if needed
        if (argc >= 2)
        {
            if (! freopen (argv [2], "w", stdout))
            {
                fprintf (stderr, "Cannot write result to file %s\n", argv [2]);
                return (2);
            }
        }
    }

    lastchar = getchar ();
    putchar ('[');

    while (lastchar != EOF)
    {
        // reading and copying mnemonic inside [... ]
        // it works also with several blank lines
        while (! isblank (lastchar) && lastchar != EOF)
        {
            // no char to escape may be in a mnemonic, but it is more sure
            // to test that and it can be usefull for comment lines
            if (char_to_escape (lastchar))
                putchar ('\\');

            putchar (lastchar);
            lastchar = getchar ();
        }

        // reading and copying spaces before the printf formatted string
        while (isblank (lastchar))
        {
            putchar (lastchar);
            lastchar = getchar ();
        }

        // normally the printf formated string begins with a "
        // but there can be comments here instead
        if (lastchar == '"')
        {
            putchar (lastchar);
            lastchar = getchar ();
        }

        // if the printf formatted string starts by \n \r or another
        // this special character will not be translated
        while (lastchar == '\\')
        {
            putchar (lastchar);
            putchar (lastchar);     // 2 '\' instead of one needed
            lastchar = getchar ();
            putchar (lastchar);
            lastchar = getchar ();
        }

        // now, the end of the line will be translated
        putchar (']');

        while (lastchar != '\r' && lastchar != '\n' && lastchar != EOF)
        {
            // except for \? sequences
            if (lastchar == '\\')
            {
                putchar ('[');

                do
                {
                    putchar (lastchar);
                    putchar (lastchar);     // 2 '\' instead of one needed
                    lastchar = getchar ();
                    putchar (lastchar);
                    lastchar = getchar ();
                }
                while (lastchar == '\\');

                putchar (']');
            }

            // and % sequences
            else if (lastchar == '%')
            {
                putchar ('[');

                do
                {
                    putchar (lastchar);
                    lastchar = getchar ();
                }
                while (isdigit (lastchar) || lastchar == '.');

                putchar (lastchar);
                putchar (']');
                lastchar = getchar ();
            }
            // but the rest is translated
            else
            {
                // special char must be escaped
                if (char_to_escape (lastchar))
                    putchar ('\\');

                putchar (lastchar);
                lastchar = getchar ();
            }
        }

        // added to say each formatted string must be translated
        // without taking into account previous formatted strings
        printf (",[]");

        // end of the formatted string reached
        putchar ('[');

        while (lastchar != '\n' && lastchar != EOF)
        {
            putchar (lastchar);
            lastchar = getchar ();
        }
    }

    // end of file
    putchar (']');
    return (0);
}
